src/SAML2/SOAP.php

```diff
@@ -18,17 +19,24 @@ class SOAP extends Binding
      */
     public function send(Message $message)
     {
-        header('Content-Type: text/xml', true);
-        $outputFromIdp = '<?xml version="1.0" encoding="UTF-8"?>';
-        $outputFromIdp .= '<SOAP-ENV:Envelope xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/">';
-        $outputFromIdp .= '<SOAP-ENV:Body>';
-        $xmlMessage = $message->toSignedXML();
-        Utils::getContainer()->debugMessage($xmlMessage, 'out');
-        $tempOutputFromIdp = $xmlMessage->ownerDocument->saveXML($xmlMessage);
-        $outputFromIdp .= $tempOutputFromIdp;
-        $outputFromIdp .= '</SOAP-ENV:Body>';
-        $outputFromIdp .= '</SOAP-ENV:Envelope>';
-        print($outputFromIdp);
+        
+        $envelope = '<soap-env:Envelope xmlns:soap-env="http://schemas.xmlsoap.org/soap/envelope/">'.
+                    '<soap-env:Header/><soap-env:Body /></soap-env:Envelope>';
+
+        $doc = new DOMDocument();
+        $doc->loadXML($envelope);
+        $soapHeader = $doc->getElementsByTagNameNS('http://schemas.xmlsoap.org/soap/envelope/', 'Header');
+        $soapBody = $doc->getElementsByTagNameNS('http://schemas.xmlsoap.org/soap/envelope/', 'Body');
+
+        if ($message->toSignedXML()->localName === 'Response') {
```

Comment:
A safer alternative may be to copy `SOAP.php` to `PAOS.php`, leaving the original alone, and removing [this hack in simplesamlphp](https://github.com/simplesamlphp/simplesamlphp/pull/233/files#diff-456d2e8c8b05c3e68f3ddf81766b1326R252).